### PR TITLE
feat: new Message Store connection listener

### PR DESCRIPTION
**Pull request by salvatore-coppola**
In order to add to the [DataServiceImpl](kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java) the capability to track the status of the used [MessageStore](kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStore.java) we added the new [ConnectionListener](kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/store/listener/ConnectionListener.java) interface.

The [MessageStoreProvider](kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStoreProvider.java) now supports the addition of a set of ConnectionListener.

### CODE REVIEW

File: kura/org.eclipse.kura.api/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -40,7 +40,7 @@ Export-Package: org.eclipse.kura;version="1.7.0",
  org.eclipse.kura.data;version="1.1.2",
  org.eclipse.kura.data.listener;version="1.0.1",
  org.eclipse.kura.data.transport.listener;version="1.0.1",
- org.eclipse.kura.db;version="2.0.0",
+ org.eclipse.kura.db;version="2.1.0",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.api/META-INF/MANIFEST.MF]: There should be no need for this change

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStoreProvider.java [hunk 2]
```diff
@@ -27,9 +28,30 @@ public interface MessageStoreProvider {
...
+
+    /**
+     * Adds a {@link ConnectionListener}
+     *
+     * @param listener
+     *            to add
+     *
+     * @since 2.5.0
+     */
+    public void addListener(ConnectionListener listener);
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStoreProvider.java:49]: We should document that the listener should close any currently open MessageStore instances when the disconnected() method is called. Can the listener do that during the disconnected() call or that operation must be done in a different thread?

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/store/listener/ConnectionListener.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ *  Eurotech
+ *******************************************************************************/
+
+package org.eclipse.kura.store.listener;
+
+import org.osgi.annotation.versioning.ConsumerType;
+
+/**
+ * Listener interface to be implemented by applications that needs to be notified about connection events in the
+ * {@link MessageStore}.
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/store/listener/ConnectionListener.java]: The MessageStore is linked here, but actually this listener can be registered in MessageStoreProvider and WireRecordStoreProvider

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/wire/store/provider/WireRecordStoreProvider.java [hunk 4]
```diff
@@ -27,10 +28,30 @@ public interface WireRecordStoreProvider {
...
 
+    /**
+     * Adds a {@link ConnectionListener}
+     *
+     * @param listener
+     *            to add
+     *
+     * @since 2.5.0
+     */
+    public void addListener(ConnectionListener listener);
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/wire/store/provider/WireRecordStoreProvider.java:50]: We should document that the listener should close any currently open WireRecordStore instance, the same note as the MessageStoreProvider.

File: kura/org.eclipse.kura.core/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -37,19 +37,21 @@ Import-Package: javax.crypto,
  org.eclipse.kura.data;version="[1.1,2.0)",
  org.eclipse.kura.data.listener;version="[1.0,2.0)",
  org.eclipse.kura.data.transport.listener;version="[1.0,2.0)",
- org.eclipse.kura.db;version="[2.0,2.1)",
+ org.eclipse.kura.db;version="[2.1,2.2)",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.core/META-INF/MANIFEST.MF]: There should be no need for this change

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/store/MessageStoreState.java [hunk 6]
```diff
@@ -29,12 +30,16 @@ public class MessageStoreState {
...
 
     private final MessageStoreProvider messageStoreProvider;
+
     private DataServiceOptions options;
     private Optional<MessageStore> messageStore = Optional.empty();
     private Optional<ScheduledExecutorService> houseKeeperExecutor = Optional.empty();
 
-    public MessageStoreState(final MessageStoreProvider messageStoreProvider, final DataServiceOptions options) {
+    public MessageStoreState(final MessageStoreProvider messageStoreProvider,
+            final ConnectionListener connectionListener, final DataServiceOptions options) {
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/store/MessageStoreState.java]: Is the connectionListener argument unused?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 7]
```diff
@@ -571,52 +637,86 @@ public int publish(String topic, byte[] payload, int qos, boolean retain, int pr
...
+                    // Priority 1 are used for remove management by Cloudlet applications.
+                    // For those messages, bypass the maximum message count check of the DB cache.
+                    // We want to publish those message even if the DB is full, so allow their
+                    // storage.
+                    if (priority != 0 && priority != 1) {
+                        int count = currentStore.getMessageCount();
+                        logger.debug("Store message count: {}", count);
+                        if (count >= this.dataServiceOptions.getStoreCapacity()) {
+                            logger.error("Store capacity exceeded");
+                            throw new KuraStoreCapacityReachedException("Store capacity exceeded");
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:665]: This one should not trigger disconnectDataTransportAndLog().

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/wire/store/AbstractJdbcWireRecordStoreImpl.java [hunk 8]
```diff
@@ -48,6 +50,8 @@ public abstract class AbstractJdbcWireRecordStoreImpl implements WireRecordStore
     protected final ConnectionProvider connectionProvider;
     protected final JdbcWireRecordStoreQueries queries;
 
+    private Set<ConnectionListener> connectionListeners;
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/wire/store/AbstractJdbcWireRecordStoreImpl.java:53]: This is unused, the changes to this file can be rolled back probably

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/db/H2DbWireRecordStoreImpl.java [hunk 9]
```diff
@@ -81,4 +91,8 @@ private static Map<Class<? extends TypedValue<?>>, String> buildTypeMapping() {
         return Collections.unmodifiableMap(result);
     }
 
+    private Optional<String> getJdbcType(final TypedValue<?> value) {
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/db/H2DbWireRecordStoreImpl.java]: This appears to be unused

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/db/H2DbWireRecordStoreImpl.java [hunk 10]
```diff
@@ -35,6 +37,14 @@ public class H2DbWireRecordStoreImpl extends AbstractJdbcWireRecordStoreImpl {
 
     private static final Map<Class<? extends TypedValue<?>>, String> TYPE_MAPPING = buildTypeMapping();
 
+    public H2DbWireRecordStoreImpl(ConnectionProvider provider, String tableName, Set<ConnectionListener> listeners)
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/db/H2DbWireRecordStoreImpl.java]: There should be no longer need for this constructor

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStore.java [hunk 11]
```diff
@@ -37,23 +37,28 @@ public interface MessageStore {
      * the <code>createdOn</code> message parameter to the current time.
      * <br>
      * 
-     * @param topic    the value of the <code>topic</code> parameter.
-     * @param payload  topic the value of the <code>payload</code> parameter.
-     * @param qos      topic the value of the <code>QoS</code> parameter.
-     * @param retain   topic the value of the <code>retain</code> parameter.
-     * @param priority topic the value of the <code>priority</code> parameter.
+     * @param topic
+     *            the value of the <code>topic</code> parameter.
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStore.java]: It looks like this file has only formatting related changes, if this is true can you please roll them back?

File: kura/org.eclipse.kura.db.sqlite.provider/META-INF/MANIFEST.MF [hunk 12]
```diff
@@ -14,13 +14,15 @@ Import-Package: com.zaxxer.hikari;version="2.7.9",
  org.eclipse.kura.configuration;version="[1.2,2.0)",
  org.eclipse.kura.crypto;version="1.3.0",
  org.eclipse.kura.data;version="[1.1,2.0)",
- org.eclipse.kura.db;version="[2.0,2.1)",
+ org.eclipse.kura.db;version="[2.1,2.2)",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.db.sqlite.provider/META-INF/MANIFEST.MF]: There should be no need for this

File: kura/org.eclipse.kura.db.sqlite.provider/META-INF/MANIFEST.MF [hunk 13]
```diff
@@ -14,13 +14,15 @@ Import-Package: com.zaxxer.hikari;version="2.7.9",
...
  org.eclipse.kura.message.store;version="[1.0,2.0)",
  org.eclipse.kura.message.store.provider;version="[1.0,1.1)",
+ org.eclipse.kura.store.listener;version="1.0.0",
  org.eclipse.kura.type;version="[1.1,2.0)",
  org.eclipse.kura.util.configuration;version="[1.0,1.1)",
  org.eclipse.kura.util.jdbc;version="[1.0,2.0)",
- org.eclipse.kura.util.message.store;version="[1.0,1.1)",
+ org.eclipse.kura.util.message.store;version="[1.0,2.0)",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.db.sqlite.provider/META-INF/MANIFEST.MF:24]: What is the reason for this change?

File: kura/org.eclipse.kura.util/META-INF/MANIFEST.MF [hunk 14]
```diff
@@ -27,6 +28,7 @@ Export-Package: org.eclipse.kura.util.base;version="1.2.0",
  org.eclipse.kura.util.message.store;version="1.0.0";x-internal:=true,
  org.eclipse.kura.util.osgi;version="1.0.0",
  org.eclipse.kura.util.service;version="1.1.0",
+ org.eclipse.kura.util.store.listener;version="1.0.0",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.util/META-INF/MANIFEST.MF]: Please mark this as x-internal

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/AbstractJdbcMessageStoreImpl.java [hunk 15]
```diff
@@ -103,8 +103,7 @@ protected long storeInternal(String topic, byte[] payload, int qos, boolean reta
 
             final long result;
 
-            try (PreparedStatement pstmt = c.prepareStatement(this.queries.getSqlStore(),
-                    new String[] { "id" })) {
+            try (PreparedStatement pstmt = c.prepareStatement(this.queries.getSqlStore(), new String[] { "id" })) {
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/AbstractJdbcMessageStoreImpl.java]: It looks like this file has only formatting related changes, if this is true can you please roll them back?

**Comment by nicolatimeus** [comment]
Build is failing

**Comment by MMaiero** [comment]
Build is now passing

**Review by nicolatimeus — APPROVED** [review]
